File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsTextLink.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 13474:16003 — mirrors web/TextLink.tsx + lib/ilds_text_link.dart.

enum class IldsTextLinkSize { Small, Medium, Large }
enum class IldsTextLinkColour { Default, White }

@Composable
fun IldsTextLink(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    size: IldsTextLinkSize = IldsTextLinkSize.Medium,
    colour: IldsTextLinkColour = IldsTextLinkColour.Default,
    isVisited: Boolean = false,
    isDisabled: Boolean = false,
    prefixIcon: (@Composable () -> Unit)? = null,
    suffixIcon: (@Composable () -> Unit)? = null,
) {
    val metrics = remember(size) { IldsTextLinkMetrics(size) }
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val colors = remember(colour, isVisited, isDisabled, isPressed) {
        IldsTextLinkColors.resolve(colour, isVisited, isDisabled, isPressed)
    }

    TextButton(
        onClick = onClick,
        modifier = modifier.semantics { role = Role.Button },
        enabled = !isDisabled,
        interactionSource = interactionSource,
    ) {
        Row(horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp4)) {
            if (prefixIcon != null) {
                Box(modifier = Modifier.semantics(mergeDescendants = true) {}) {
                    prefixIcon()
                }
            }
            Text(
                text = label,
                style = metrics.textStyle.copy(
                    color = colors.text,
                    textDecoration = if (isDisabled) TextDecoration.None else TextDecoration.Underline,
                ),
            )
            if (suffixIcon != null) {
                Box(modifier = Modifier.semantics(mergeDescendants = true) {}) {
                    suffixIcon()
                }
            }
        }
    }
}

@Immutable
private data class IldsTextLinkColors(val text: Color) {
    companion object {
        fun resolve(
            colour: IldsTextLinkColour,
            isVisited: Boolean,
            isDisabled: Boolean,
            isPressed: Boolean,
        ): IldsTextLinkColors {
            if (colour == IldsTextLinkColour.White) {
                val c = when {
                    isDisabled -> IldsTokens.neutralCoolgray400
                    isVisited -> IldsTokens.neutralCoolgray300
                    isPressed -> IldsTokens.neutralCoolgray300
                    else -> IldsTokens.globalWhite000
                }
                return IldsTextLinkColors(c)
            }
            val c = when {
                isDisabled -> IldsTokens.neutralCoolgray300
                isVisited -> IldsTokens.neutralCoolgray500
                isPressed -> IldsTokens.informativeBlue700
                else -> IldsTokens.informativeBlue500
            }
            return IldsTextLinkColors(c)
        }
    }
}

@Immutable
private data class IldsTextLinkMetrics(val textStyle: TextStyle) {
    constructor(size: IldsTextLinkSize) : this(
        textStyle = TextStyle(
            fontSize = when (size) {
                IldsTextLinkSize.Small -> IldsTokens.fontSize12
                IldsTextLinkSize.Medium -> IldsTokens.fontSize14
                IldsTextLinkSize.Large -> IldsTokens.fontSize16
            },
            lineHeight = when (size) {
                IldsTextLinkSize.Small -> 16.sp
                IldsTextLinkSize.Medium -> 18.sp
                IldsTextLinkSize.Large -> 20.sp
            },
            fontWeight = IldsTokens.fontWeightMedium,
        ),
    )
}

File: ios/Sources/ILDSDesignSystem/IldsTextLink.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_text_link.dart + web TextLink.tsx. TODO(web/desktop): hover states — mobile-first defer.

public enum IldsTextLinkSize: Sendable {
    case small, medium, large
}

public enum IldsTextLinkColour: Sendable {
    case defaultBlue, white
}

public struct IldsTextLink: View {
    private let label: String
    private let action: () -> Void
    private let size: IldsTextLinkSize
    private let colour: IldsTextLinkColour
    private let isVisited: Bool
    private let isDisabled: Bool
    private let prefixIcon: AnyView?
    private let suffixIcon: AnyView?

    public init(
        _ label: String,
        action: @escaping () -> Void,
        size: IldsTextLinkSize = .medium,
        colour: IldsTextLinkColour = .defaultBlue,
        isVisited: Bool = false,
        isDisabled: Bool = false,
        prefixIcon: (any View)? = nil,
        suffixIcon: (any View)? = nil
    ) {
        self.label = label
        self.action = action
        self.size = size
        self.colour = colour
        self.isVisited = isVisited
        self.isDisabled = isDisabled
        self.prefixIcon = prefixIcon.map { AnyView($0) }
        self.suffixIcon = suffixIcon.map { AnyView($0) }
    }

    public var body: some View {
        Button(action: action) {
            IldsTextLinkLabel(
                label: label,
                size: size,
                colour: colour,
                isVisited: isVisited,
                isDisabled: isDisabled,
                isPressed: false,
                prefixIcon: prefixIcon,
                suffixIcon: suffixIcon
            )
        }
        .buttonStyle(IldsTextLinkButtonStyle(
            label: label,
            size: size,
            colour: colour,
            isVisited: isVisited,
            isDisabled: isDisabled,
            prefixIcon: prefixIcon,
            suffixIcon: suffixIcon
        ))
        .disabled(isDisabled)
        .accessibilityLabel(label)
        .accessibilityAddTraits(.isLink)
    }
}

private struct IldsTextLinkLabel: View {
    let label: String
    let size: IldsTextLinkSize
    let colour: IldsTextLinkColour
    let isVisited: Bool
    let isDisabled: Bool
    let isPressed: Bool
    let prefixIcon: AnyView?
    let suffixIcon: AnyView?

    var body: some View {
        let metrics = IldsTextLinkMetrics(size: size)
        let color = IldsTextLinkColors.resolve(
            colour: colour,
            isVisited: isVisited,
            isDisabled: isDisabled,
            isPressed: isPressed
        )

        HStack(spacing: ILDSTokens.sp4) {
            if let prefixIcon {
                prefixIcon
                    .foregroundStyle(color)
                    .frame(width: metrics.iconSize, height: metrics.iconSize)
            }
            Text(label)
                .font(metrics.font)
                .foregroundStyle(color)
                .underline(!isDisabled)
            if let suffixIcon {
                suffixIcon
                    .foregroundStyle(color)
                    .frame(width: metrics.iconSize, height: metrics.iconSize)
            }
        }
    }
}

private struct IldsTextLinkButtonStyle: ButtonStyle {
    let label: String
    let size: IldsTextLinkSize
    let colour: IldsTextLinkColour
    let isVisited: Bool
    let isDisabled: Bool
    let prefixIcon: AnyView?
    let suffixIcon: AnyView?

    func makeBody(configuration: Configuration) -> some View {
        IldsTextLinkLabel(
            label: label,
            size: size,
            colour: colour,
            isVisited: isVisited,
            isDisabled: isDisabled,
            isPressed: configuration.isPressed && !isDisabled,
            prefixIcon: prefixIcon,
            suffixIcon: suffixIcon
        )
    }
}

private struct IldsTextLinkMetrics {
    let font: Font
    let iconSize: CGFloat

    init(size: IldsTextLinkSize) {
        switch size {
        case .small:
            font = .system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize12
        case .medium:
            font = .system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize14
        case .large:
            font = .system(size: ILDSTokens.fontSize16, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize16
        }
    }
}

private struct IldsTextLinkColors {
    static func resolve(
        colour: IldsTextLinkColour,
        isVisited: Bool,
        isDisabled: Bool,
        isPressed: Bool
    ) -> Color {
        if colour == .defaultBlue {
            if isDisabled { return ILDSTokens.neutralCoolgray300 }
            if isVisited { return ILDSTokens.neutralCoolgray500 }
            if isPressed { return ILDSTokens.informativeBlue700 }
            return ILDSTokens.informativeBlue500
        }
        if isDisabled { return ILDSTokens.neutralCoolgray400 }
        if isVisited { return ILDSTokens.neutralCoolgray300 }
        if isPressed { return ILDSTokens.neutralCoolgray300 }
        return ILDSTokens.globalWhite000
    }
}
